pkg/lib/oauth/handler/handler_token.go

```diff
@@ -207,8 +215,28 @@ func (h *TokenHandler) handleAuthorizationCode(
 		return nil, protocol.NewError("invalid_request", "invalid redirect URI")
 	}
 
-	if codeGrant.PKCEChallenge != "" && !verifyPKCE(codeGrant.PKCEChallenge, r.CodeVerifier()) {
-		return nil, errInvalidAuthzCode
+	// if code verifier is provided, must verify PKCE
+	if r.CodeVerifier() != "" {
+		if codeGrant.PKCEChallenge == "" || !verifyPKCE(codeGrant.PKCEChallenge, r.CodeVerifier()) {
+			return nil, errInvalidAuthzCode
+		}
+	}
+
+	// verify client secret
+	if r.ClientSecret() != "" {
+		credentialsItem, ok := h.OAuthClientCredentials.Lookup(client.ClientID)
+		if !ok {
+			return nil, protocol.NewError("invalid_request", "client secret is not supported for the client")
+		}
+		key, _ := credentialsItem.Set.Get(0)
+		var expectedClientSecret []byte
+		err = key.Raw(&expectedClientSecret)
+		if err != nil {
+			return nil, err
+		}
+		if subtle.ConstantTimeCompare([]byte(r.ClientSecret()), []byte(expectedClientSecret)) != 1 {
```

Comment:
We will show `k` to the developer, which is a base64URL encoded string. `key.Raw` will decode `k` into bytes. So they are not going to match.

Reply:
Per our discussion. Since the decoded oct key is printable, we are going to use the decoded value as the client secret, like the webhook secret key.